tests/0085-headers.cpp

```diff
@@ -0,0 +1,219 @@
+
+  std::vector<RdKafka::Headers::Header> headers_arr;
+  for (int i = 0; i < n; ++i) {
+    std::stringstream key_s;
+    key_s << "header_" << i;
+    std::string key = key_s.str();
+    std::stringstream val_s;
+    val_s << "value_" << i;
+    std::string val = val_s.str();
+    headers_arr.push_back(RdKafka::Headers::Header(key, val.c_str()));
+  }
+  RdKafka::Headers *produce_headers = RdKafka::Headers::create(headers_arr, false);
+
+  err = p->produce(topic, 0,
+                    RdKafka::Producer::RK_MSG_COPY,
+                    (void *)message, message ? strlen(message) : 0,
+                    (void *)"key", 3, 0, NULL, produce_headers);
+
+  p->flush(tmout_multip(10000));
+
+  if (p->outq_len() > 0)
+    Test::Fail(tostr() << "Expected producer to be flushed, " <<
+               p->outq_len() << " messages remain");
+
+  RdKafka::KafkaConsumer *c = RdKafka::KafkaConsumer::create(conf, errstr);
+  if (!c)
+    Test::Fail("Failed to create KafkaConsumer: " + errstr);
+  
+  std::vector<std::string> topics;
+  topics.push_back(topic);
+  if ((err = c->subscribe(topics)))
+    Test::Fail("subscribe failed: " + RdKafka::err2str(err));
+
+  int cnt = 0;
+  while (!cbs.seen_commit || !cbs.seen_stats) {
+    RdKafka::Message *msg = c->consume(tmout_multip(1000));
+    if (!msg->err()) {
+      cnt++;
+      Test::Say(tostr() << "Received message #" << cnt << "\n");
+      if (cnt > 10)
+        Test::Fail(tostr() << "Should've seen the "
+                   "offset commit (" << cbs.seen_commit << ") and "
+                   "stats callbacks (" << cbs.seen_stats << ") by now");
+
+      /* Commit the first message to trigger the offset commit_cb */
+      if (cnt == 1) {
+        err = c->commitAsync(msg);
+        if (err)
+          Test::Fail("commitAsync() failed: " + RdKafka::err2str(err));
+        rd_sleep(1); /* Sleep to simulate slow processing, making sure
+                      * that the offset commit callback op gets
+                      * inserted on the consume queue in front of
+                      * the messages. */
+      }
+      RdKafka::Headers *headers = msg->get_headers();
+      if (!headers) {
+        Test::Fail("Expected RdKafka::Message to contain headers");
+      }
+
+      assert_all_headers_match(headers, headers_arr);
+
+    } else if (msg->err() == RdKafka::ERR__TIMED_OUT)
+      ; /* Stil rebalancing? */
+    else
+      Test::Fail("consume() failed: " + msg->errstr());
+    delete msg;
+  }
+
+  c->close();
+  delete c;
+  delete p;
+  delete conf;
+
+}
+
+static void test_one_header () {
+  Test::Say("Test one header in consumed message.\n");
+  std::string val = "valid";
+  test_n_headers(1, val.c_str());
+}
+
+static void test_ten_headers () {
+  Test::Say("Test ten headers in consumed message.\n");
+  std::string val = "valid";
+  test_n_headers(10, val.c_str());
+}
+
+static void test_one_header_null_msg () {
+  Test::Say("Test one header in consumed message with a null value message.\n");
```

Comment:
I don't think this very valuable, the message value / key are not really related to the headers

Reply:
Just on this one when I produce messages with null or empty values the headers do not come through on the consume, is this expected behaviour?